pkg/uploader/kopia/block_backup.go

```diff
@@ -0,0 +1,52 @@
+/*
+Copyright The Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kopia
+
+import (
+	"os"
+	"syscall"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/kopia/kopia/fs/virtualfs"
+	"github.com/pkg/errors"
+)
+
+const ErrNotPermitted = "operation not permitted"
+
+func getLocalBlockEntry(kopiaEntry fs.Entry) (fs.Entry, error) {
+	path := kopiaEntry.LocalFilesystemPath()
+
+	fileInfo, err := os.Lstat(path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get the source device information %s", path)
+	}
+
+	if (fileInfo.Sys().(*syscall.Stat_t).Mode & syscall.S_IFMT) != syscall.S_IFBLK {
+		return nil, errors.Errorf("source path %s is not a block device", path)
+	}
+
+	device, err := os.Open(path)
+	if err != nil {
+		if os.IsPermission(err) || err.Error() == ErrNotPermitted {
+			return nil, errors.Wrapf(err, "no permission to open the source device %s, make sure that node agent is running in privileged mode", path)
```

Comment:
If we need to add the explanation because it is not clear why in the design, that makes sense to me, just noting that we did mention the need for the privileged access and mounting the other location.

Reply:
My understanding and @dzaninovic can correct me if I am wrong, but we need to have root to have access to the devices that are symlinked in the plugins directory. 

It sounds to me not that we are in disagreement that you need to have this node agent run as root, but rather how this is set up. 

In the PR, we add the configuration to the node agent on the velero install 

You are suggesting documentation that says, to use this, you need to have privileged mode for the node agent, and here is how you set it up. Is my understanding correct?

Either way, this error message is probably useful right?

Reply:
Root user and privileged mode are different, node-agent has already been running as root user, but it doesn't run under privileged mode except for some special envs.
Privileged mode brings more security risks, so users generally don't want it. This is why I say if we can remove this constraint, we should remove it, otherwise, we should document it as a limitation.

Reply:
Sorry, I was using them interchangeably, and they are not you are correct. 

I tested it on Openshift, and it will need to be privileged there (I tested without and did get permission errors when accessing the device). 

IIRC @dzaninovic and I also got it working on minikube, and it also needed to have privileged mode in that env. I would think, that adding the need for the user to add this and set it up correctly to use block device would be fine if this is a blocker. 

@dzaninovic can you confirm this is correct?

Reply:
I think for OpenShift, privileged mode is required to access host path, see `[some example here](https://github.com/vmware-tanzu/velero/blob/main/site/content/docs/main/file-system-backup.md#:~:text=To%20mount%20the%20correct%20hostpath%20to%20pods%20volumes%2C%20run%20the%20node%2Dagent%20pod%20in%20privileged%20mode.)` , because OpenShift is running under SELINUX.
However, for most other envs, privileged mode is not required.
So if accessing host path is the only consideration for privileged mode, I think we can remove it, specifically:
```
Remove it from the node-agent daemonset spec
Remove the installation flag
Remove the check and error here
```

If privileged mode is required for any other reason, please share the details.

Reply:
Yes, we need the privileged mode to open the device on Minikube and that is why I added --privileged-agent option to the Velero installer.

Agent container can't access the device as it does not have the device attached to it.
Exposer container would be able to access the device without the privileged mode so to avoid using it we would need to run a backup from there.  This is what we do in CloudCasa.

Reply:
Yes, we could use the already existing generic solution as you suggested and remove the flag.
Flag is making it easier to configure the privileged mode so if privileged mode is always required do we want to keep the flag?
I could not find anything in the Kubernetes documentation about privileged mode being required for this case but it would be good to test on at least one configuration where privileged mode is not required to confirm that such configurations exist.
If @shawn-hurley finds time he will test this on DigitalOcean cluster, if not I will try to test it on EKS or DigitalOcean.

Reply:
Thanks for the testing. As far as I know, these envs don't need privileged mode to access host paths -- EKS, AKS, GCP.

Reply:
> Flag is making it easier to configure the privileged mode so if privileged mode is always required do we want to keep the flag?

Whether to add this flag, let's discuss under [this comment](https://github.com/vmware-tanzu/velero/pull/6680#discussion_r1320940767), If host path access is the only requirement, just one thing:

- If host path access is the only requirement, then it is the existing general requirement for all data mover cases, not only for block level backup.
- Then if we want to add this flag, we add it for the entire CSI data mover feature, as well as fs backup feature. So we need to change the both document
- Then it is better to create a new PR, as the current PR is already large and this change is not for block level backup

Reply:
If host path access is the only requirement, we need to remove the below statement from the design, it is confusing --- when I first noticed this, I thought the privileged mode is required all the time by block level backup for some special reasons. 
Moreover, the previous sections in the designed doc has already clarified the entire data mover design require node-agent to run under root user, so we don't need to mention root user either.

> Of note, we do need to add root access to the daemon set node agent to access the new mount. 
> ```yaml
> ...
>             - mountPath: /var/lib/kubelet/plugins
>               mountPropagation: HostToContainer
>               name: host-plugins
> ....
>         - hostPath:
>             path: /var/lib/kubelet/plugins
>           name: host-plugins
> ...
> 							SecurityContext: &corev1.SecurityContext{
> 								Privileged: &c.privilegedAgent,
> 							},
> ```

Reply:
> Thanks for the testing. As far as I know, these envs don't need privileged mode to access host paths -- EKS, AKS, GCP, vanilla clusters.

Even on MiniKube I don't need privileged mode to access the mounted filesystem but I need it to access the device so it is possible that above clusters will also need it.  I will test it with EKS.

Reply:
@Lyndon-Li I did double-check on EKS with the ebs CSI driver. 

I am getting permission errors. As David mentioned, I have a feeling that this has to do with opening the device on the node (as that is the underlying location), which needs host-level permission, not just root. Testing this on all platforms to prove it is the case for everything though is not possible.

```
time="2023-09-18T17:13:20Z" level=warning msg="Failed to patch dataupload with err Failed to run kopia backup: unable to get local block device entry: no permission to open the source device /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-658e0d23-3803-4d08-ae1a-169ff62c3d33/c2936d79-1c65-42cb-ba6a-59e0b74ccff7, make sure that node agent is running in privileged mode: open /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-658e0d23-3803-4d08-ae1a-169ff62c3d33/c2936d79-1c65-42cb-ba6a-59e0b74ccff7: operation not permitted" dataupload=shawn-1-kv9j7 error="Failed to run kopia backup: unable to get local block device entry: no permission to open the source device /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-658e0d23-3803-4d08-ae1a-169ff62c3d33/c2936d79-1c65-42cb-ba6a-59e0b74ccff7, make sure that node agent is running in privileged mode: open /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-658e0d23-3803-4d08-ae1a-169ff62c3d33/c2936d79-1c65-42cb-ba6a-59e0b74ccff7: operation not permitted" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/uploader/kopia/block_backup.go:45" error.function=github.com/vmware-tanzu/velero/pkg/uploader/kopia.getLocalBlockEntry logSource="pkg/controller/data_upload_controller.go:407"
```

At this point, I would like to make a compromise. 

1. We remove the option for installation. as we don't know for a fact if it is needed for everything. If we find that users must do it too often on their own, we can add it back later.
2. I think keeping the error context and checking makes it very clear what is happening. Because this happened on EKS and I would have to update the install, I would really think for the UX of this, that we keep it. 

What do folks think?

Reply:
@shawn-hurley 
Thanks for the testing.
Personally, I think we don't need to prove this in all platforms. Once we approve that the privileged mode is required in the popular platforms, we can keep the current check and it's better to keep the installation flag as well.

Then following this, we need to do some more things:

1. **We should add the clarification in both data mover design and document that block level backup requires privileged mode and a short explanation of why it is required in the design**. Otherwise, some time later, people would doubt if it really a hard requirement and do lots of the same verification again.
2. Since the fs-backup and the general data mover also needs privileged mode in some cases and we have documented to manual configure the privileged mode. If we decide to add the installation flag, the fs-backup and general data mover's better to use the same flag, so **the two documents should be changed --- the manual way should be replaced**.

Reply:
I modified the design document to explain why is additional mount and privileged mode required.  User documentation still needs to be modified.

Commit: https://github.com/vmware-tanzu/velero/pull/6680/commits/1ae3e6cf9c83acb52e803adca1b8e11f439cd700

Reply:
> 2. Since the fs-backup and the general data mover also needs privileged mode in some cases and we have documented to manual configure the privileged mode. If we decide to add the installation flag, the fs-backup and general data mover's better to use the same flag, so **the two documents should be changed --- the manual way should be replaced**.

I confirmed on DigitalOcean that privileged mode is required:
```
time="2023-09-25T15:52:35Z" level=error msg="Async fs backup data path failed" dataupload=backup1-vnw92 error="Failed to run kopia backup: unable to get local block device entry: no permission to open the source device /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-447a6f0a-5db4-4087-acfd-4d593355a268/e4e21946-64fc-4efc-84cc-398a7509710b, make sure that node agent is running in privileged mode: open /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-447a6f0a-5db4-4087-acfd-4d593355a268/e4e21946-64fc-4efc-84cc-398a7509710b: operation not permitted" error.file="/go/src/github.com/vmware-tanzu/velero/pkg/uploader/kopia/block_backup.go:48" error.function=github.com/vmware-tanzu/velero/pkg/uploader/kopia.getLocalBlockEntry logSource="pkg/controller/data_upload_controller.go:400"
```

So far we know that Minikube, EKS and DigitalOcean require privileged mode.
@Lyndon-Li, which other platforms do we need to test to confirm that we want to keep the "privileged-agent" installer option?

Reply:
@dzaninovic 
Thanks for the tests. I don't think we need any more tests, we can prove the privileged mode is required by block mode backup now.

For the installation flag, personally I think it is better to have. Let's others' opinions.

Additionally, see the [some things more] we need to do around this in comment https://github.com/vmware-tanzu/velero/pull/6680#discussion_r1329556381.

Reply:
This should be resolved with the latest commit.